### PR TITLE
 #106 Early version of test (required columns, json)

### PR DESCRIPTION
**Pull request by Ciheim**
This an early version of our testing infrastructure. This version looks for incorrect json modifications and empty values within required columns.

### CODE REVIEW

**Comment by Ciheim** [comment]
The script takes two arguments, (1) The path to the JSON that was generated by the catalog builder and (2)  The path to the JSON template that was used for generation.

It is called using this format: test.py <custom json> <json template>

The script compares the json and returns any json sections that have changes (besides the catalog path section which we expect to change).

The script also grabs all required columns, as defined by the groupby_attrs section,  and checks for any null values within them.

Here's an example I ran: `test.py ~/output.json ../cats/gfdl_template.json`

And here's the output:  
```attributes section of JSON does not refect template
source_id contains empty values
member_id contains empty values```

**Comment by ceblanton** [comment]
The test is failing on     `from jsondiff import diff`. 

meta.yaml needs an update to install jsondiff.

**Comment by Ciheim** [comment]
> The test is failing on `from jsondiff import diff`.
> 
> meta.yaml needs an update to install jsondiff.

I made the change to environment.yml in #126 but you're right. I forgot to add it to meta.yaml. I'll remove the jsondiff change there and add both changes to this PR. Nice Catch.

**Comment by ceblanton** [comment]
what is environment.yaml used for? Almost too many yamls in the world (just kidding..)

**Comment by Ciheim** [comment]
> what is environment.yaml used for? Almost too many yamls in the world (just kidding..)

We expect users who want to run the tool directly to create an environment from this yaml

**Comment by ceblanton** [comment]
Hello, I tested this and think it is good and ready. I rephrased the error messages slightly and rephrased the try/except logic.

One thing I have mixed feelings about is that it gives the bad news in a drip-drip way rather than giving all the bad news at once.  I think it's ok though as the problems are likely to be repeated.  It would probably be worse to give the same error 1000x times for a 1000 line catalog.

**Review by ceblanton — APPROVED** [review]

**Comment by Ciheim** [comment]
> Hello, I tested this and think it is good and ready. I rephrased the error messages slightly and rephrased the try/except logic.
> 
> One thing I have mixed feelings about is that it gives the bad news in a drip-drip way rather than giving all the bad news at once. I think it's ok though as the problems are likely to be repeated. It would probably be worse to give the same error 1000x times for a 1000 line catalog.

Yes you're right about this. Thanks for the changes.  Bennett and I noticed it so I put in a fix but the commit got a little messed up. Git errors bleh... I'll take care of this soon.

File: tests/test_catalog.py
Comment (aradhakrishnanGFDL) [inline tests/test_catalog.py:1]: Add docstring at the top to explain what test_catalog.py does.
Ensure there is --help to show usage info.

File: tests/test_catalog.py [hunk 1]
```diff
@@ -1,30 +1,60 @@
...
+import click
+import json
+from jsondiff import diff
+import pandas as pd
+import sys
+
+@click.command()
+@click.argument('json_path', nargs = 1 , required = True)
+@click.argument('json_template_path', nargs = 1 , required = False)
+@click.option('-tf', '--test-failure', is_flag=True, default = False, help="Errors are only printed. Program will not exit.")
```
Comment (aradhakrishnanGFDL) [inline tests/test_catalog.py:12]: document the purpose of this flag so user can choose to use or ignore.
Sounds like sys.exit is escaped when it comes to the CI usage. 
How does a successful test show up? 

Also suggest renaming tests/config.yaml to a self-described name.  
Wonder why member_id is null, we changed the code to say "na". Was this a made-up test-case?

File: tests/test_catalog.py [hunk 2]
```diff
@@ -1,30 +1,60 @@
...
+            sys.exit("Catalog has no values")
+
+    #Get required columns
+    req = (j["aggregation_control"]["groupby_attrs"])
+ 
+    #Check the csv headers for required columns/values
+    errors = 0
+    for column in req:
+        if column not in catalog.columns:
+            print(f"The required column '{column}' does not exist. Double check config file.")
```
Comment (aradhakrishnanGFDL) [inline tests/test_catalog.py:45]:
```suggestion
            print(f"The required column '{column}' does not exist in the csv. In other words, there is some inconsistency between the json and the csv file. Please check out info listed under aggregation_control and groupby_attrs in your json file and verify if those columns show up in the csv as well.")
```

File: tests/test_catalog.py (hunk 1 above)
Comment (Ciheim) [reply tests/test_catalog.py:12]: Where do you think it should be documented?

Comment (Ciheim) [reply tests/test_catalog.py:12]: The test failure is a temporary solution until the test-case is fixed. Can't remember where it came from.. I copied the existing test config into the config yaml.

Comment (Ciheim) [reply tests/test_catalog.py:12]: > Where do you think it should be documented?

Nevermind.. I'll include it in the docstrings

Comment (Ciheim) [reply tests/test_catalog.py:12]: Also, a successful test shouldn't return anything.

**Review by ceblanton — APPROVED** [review]

**Pull request merged**